File: python3/merge_two_sorted_lists.py
"""
Space   : O(1)
Time    : O(n)
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        p1, p2 = l1, l2
        ans = ListNode(0)
        pans = ans
    
        while p1 and p2:
            if p1.val < p2.val:
                pans.next = ListNode(p1.val)
                p1 = p1.next
            else:
                pans.next = ListNode(p2.val)
                p2 = p2.next
            pans = pans.next
        
        if p1:
            pans.next = p1

        if p2:
            pans.next = p2
        
        return ans.next

File: golang/merge_two _sorted_lists.go
package main

import (
	"golang/model"
)

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func mergeTwoLists(l1 *model.ListNode, l2 *model.ListNode) *model.ListNode {
	p1, p2 := l1, l2
	ans := &model.ListNode{
		Val: 0,
	}
	pans := ans

	for p1 != nil && p2 != nil {
		if p1.Val < p2.Val {
			pans.Next = &model.ListNode{
				Val: p1.Val,
			}
			p1 = p1.Next
		} else {
			pans.Next = &model.ListNode{
				Val: p2.Val,
			}
			p2 = p2.Next
		}
		pans = pans.Next
	}

	if p1 != nil {
		pans.Next = p1
	}

	if p2 != nil {
		pans.Next = p2
	}

	return ans.Next
}
